ext/pdo/pdo_stmt.c

```diff
@@ -2030,7 +2031,11 @@ static HashTable *dbstmt_get_gc(zend_object *object, zval **gc_data, int *gc_cou
 
 	zend_get_gc_buffer *gc_buffer = zend_get_gc_buffer_create();
 	zend_get_gc_buffer_add_zval(gc_buffer, &stmt->database_object_handle);
-	zend_get_gc_buffer_add_zval(gc_buffer, &stmt->fetch.into);
+	if ((stmt->default_fetch_type & PDO_FETCH_INTO) == PDO_FETCH_INTO) {
+		zend_get_gc_buffer_add_obj(gc_buffer, stmt->fetch.into);
+	} else if ((stmt->default_fetch_type & PDO_FETCH_CLASS) == PDO_FETCH_CLASS) {
```

Comment:
Using a bitwise AND here seems wrong, as they're not bitmasks, so you're going to hit the condition for more than just PDO_FETCH_INTO or PDO_FETCH_CLASS.
See https://github.com/php/php-src/blob/8ed0d7fe9688c1839e1be67bda28eaae2e9fb0bc/ext/pdo/php_pdo_driver.h#L66-L81

Reply:
The problem is that the `default_fetch_type` is a union of the fetch mode, with the fetch flags. I could bitwise not it with `PDO_FETCH_FLAGS` to separate out the flags from the mode as a bunch of other places do it.

Reply:
Okay I remember now, but it's still wrong: you should AND with PDO_FETCH_FLAGS otherwise the following can happen:

Known: `PDO_FETCH_CLASS` equals 8 (0b1000), `PDO_FETCH_INTO` equals 9 (0b1001).
Problem example: let's say `default_fetch_type` is `PDO_FETCH_INTO`, and we look at the condition `(stmt->default_fetch_type & PDO_FETCH_CLASS) == PDO_FETCH_CLASS`. This condition expands to: `(0b1001 & 0b1000) == 0b1000`. We know that `0b1001 & 0b1000` is `0b1000`, so it will be equal to `PDO_FETCH_CLASS` even though that's not what we want. In this _particular_ case you don't run into it because of the ordering of the branches but that's extremely fragile and I'm sure you can come up with another combination of modes to trigger this kind of behaviour.